common.go

```diff
@@ -81,35 +84,94 @@ func detachedSignatureInputFromHash(plaintextAndHeaderHash []byte) []byte {
 	return buf.Bytes()
 }
 
-func hmacSHA512256(key []byte, input []byte) []byte {
-	// Equivalent to crypto_auth, but using Go's builtin HMAC. Truncates
-	// SHA512, instead of actually calling SHA512/256.
-	if len(key) != cryptoAuthKeyBytes {
-		panic("Bad crypto_auth key length")
+func copyEqualSize(out, in []byte) {
+	if len(out) != len(in) {
+		panic(fmt.Sprintf("len(out)=%d != len(in)=%d", len(out), len(in)))
 	}
-	authenticatorDigest := hmac.New(sha512.New, key)
-	authenticatorDigest.Write(input)
-	fullMAC := authenticatorDigest.Sum(nil)
-	return fullMAC[:cryptoAuthBytes]
+	copy(out, in)
+}
+
+func copyEqualSizeStr(out []byte, in string) {
+	if len(out) != len(in) {
+		panic(fmt.Sprintf("len(out)=%d != len(in)=%d", len(out), len(in)))
+	}
+	copy(out, in)
+}
+
+type payloadHash [sha512.Size]byte
+
+type payloadAuthenticator [sha512.Size256]byte
+
+func (pa payloadAuthenticator) Equal(other payloadAuthenticator) bool {
+	return hmac.Equal(pa[:], other[:])
+}
+
+func authenticatePayload(macKey macKey, payloadHash payloadHash) payloadAuthenticator {
+	// Equivalent to crypto_auth, but using Go's builtin HMAC.
+	authenticatorDigest := hmac.New(sha512.New512_256, macKey[:])
+	authenticatorDigest.Write(payloadHash[:])
+	var auth payloadAuthenticator
+	h := authenticatorDigest.Sum(nil)
+	copyEqualSize(auth[:], h)
+	return auth
 }
 
-func computeMACKey(secret BoxSecretKey, public BoxPublicKey, headerHash []byte) []byte {
+type macKey [cryptoAuthKeyBytes]byte
+
+func computeMACKeySingle(secret BoxSecretKey, public BoxPublicKey, headerHash headerHash) macKey {
 	nonce := nonceForMACKeyBox(headerHash)
```

Comment:
I think we're going to want to increment this nonce in V2, to avoid the problem where two recipients compute the same MAC key and then have visibly identical authenticators.